### ISSUE TITLE
Macro key options missing.

### ISSUE THREAD

**Issue by dpnova** [polychromatic/polychromatic#70]
Love that this exists! Been a long time coming.

I have an issue with my BW alternate - there's only a help button next to the macro keys option and the help it provides doesn't actually work. I've seen places where it seems the macro keys are an option in the menu etc?  http://www.omgubuntu.co.uk/2016/08/razer-chroma-keyboard-linux-configure-app

Would love to get these working!

Thanks

**Comment by lah7** [comment]
At the moment, the daemon **only** binds on-the-fly macros to the dedicated M1 to M5 keys. If the keyboard doesn't have them, this feature won't work.

I have the BW Ultimate 2016 in which macros won't record because of this restriction by the daemon. It was added to stop people "accidentally" overriding the wrong key on BW Chroma keyboards long ago.

There's no interface to manage macros yet, as the daemon doesn't support it, but it is in the process of being overhauled by @terrycain for a future feature. :+1:

**Comment by dpnova** [comment]
Yeah that matches my understanding too. Just to clarify I'm doing this:

1. press fn and (m)  together - m being the macro key or right hand alt key - at this point i don't see the macro light turn on or anything
2. press m1 the top macro key
3. press z, then a
4. repeat step 1
5. in a text doc, press m1, nothing happens.

I've done some digging into the driver device files and it seems step one just toggles the device_mode.

Have I missed something?

Two other points of note for me:
1. m3 - m5 actually emit key codes for me in lubuntu - they are some of the the x86 launch keys.
2. application profiles don't seem to save for me - not sure if that is relevant.. they appear in the applet context menu, but not the main UI - I can open another ticket for that if you think it's something to follow up.

**Comment by terrycain** [comment]
> There's no interface to manage macros yet,

lies, its there ;-)

**Comment by lah7** [comment]
What keyboard model is it? [I'd suggest raising an issue on the daemon repository](https://github.com/terrycain/razer-drivers/issues) so the driver/daemon guys can investigate. Entering the right mode should be automatic now -- a while back it use to require an "Activate" macros button.

> application profiles don't seem to save for me - not sure if that is relevant.. they appear in the applet context menu, but not the main UI - I can open another ticket for that if you think it's something to follow up.

Not sure if you mean profiles don't save at all or they simply aren't listing. Have you definitely got the latest version? (`0.3.7.1` -- this fixes a saving issue).

--------

> lies, its there ;-)

Oops, you're right. Well, one to be done here after the macro revamp is complete.

**Comment by dpnova** [comment]
Sorry what do you mean by automatic? My apologies I've not actually used the razer stuff in windows on the kb before so my overall understanding may be a bit lacking :D 

Below is package info and some screenshots.

```
➜  ~ sudo apt-cache show polychromatic
Package: polychromatic
Priority: optional
Section: x11
Installed-Size: 2092
Maintainer: Luke Horwell <luke@ubuntu-mate.org>
Architecture: all
Version: 0.3.7.1-yakkety
Recommends: fonts-play
Depends: python3 (>= 3.4.0), python3-gi, python3-setproctitle, python3-requests, gir1.2-gtk-3.0, gir1.2-webkit2-4.0 (>= 2.12.0), gir1.2-appindicator3-0.1, python3-razer, razer-daemon
Filename: pool/main/p/polychromatic/polychromatic_0.3.7.1-yakkety_all.deb
Size: 747108
MD5sum: 47eea8dd3679586c3d1aae1052bc0f69
SHA1: bb3065ec45e5b2723dc31df301c0450ce1209f62
SHA256: aa773a11202708a2b31036e0333b986d03448123e7e5f7697f580299238ee431
Description-en: Graphical front-end for managing Razer peripherals.
 An application that allows multiple Razer devices to be configured
 on the system through an easy-to-use interface.
 Requires the razer-drivers daemon.
Description-md5: 5744734ede083ab0ef8289560a719c68
```


![selection_001](https://cloud.githubusercontent.com/assets/229943/24326717/433ac2ec-1200-11e7-9827-9efe1981c51a.png)

In the below screenshot the only way I could actually get the entry to appear is by finding the json file for the profile I added previously that didnt display and adding a "name" field to it. I guess because I can't customize backlight stuff with this KB there may actually not be anything I can do with profiles, I was more checking them out to see if it fixed my macro woes.

![selection_002](https://cloud.githubusercontent.com/assets/229943/24326727/7e9b61c0-1200-11e7-8bc8-494693af50af.png)

**Comment by dpnova** [comment]
btw i'm 100% comfortable hacking on python code (and playing around in /sys) - so if you want me to go digging just point me in the right direction

**Comment by terrycain** [comment]
@lah7 would probably pay to get the output of the capabilities dict

**Comment by lah7** [comment]
Ahh, it's some sort of [BlackWidow Classic](https://www.razerzone.com/ap-en/gaming-keyboards-keypads/razer-blackwidow-ultimate-classic). That's confusing, since it's a model without any lighting (right?)

> Sorry what do you mean by automatic? My apologies I've not actually used the razer stuff in windows on the kb before so my overall understanding may be a bit lacking :D

Just something technical that is handled by the daemon. The BlackWidow Chroma keyboard for example has one of two modes AFAIK - one of them allows macros and special features (like game mode) to be used, the other is like a "generic" keyboard interface.

Feel free to hack around with the Python daemon itself *(this is what Polychromatic interfaces with)*

In the terminal: `python3`
```
import razer.client
a = razer.client.DeviceManager()
b = a.devices[0]
b.name
b.capabilities
b. <tab for functions>
```

Please do get me the output of `b.capabilities` -- I'd like to check what your device supports so Polychromatic can hide profiles for unsupported devices -- currently, it assumes if it's a keyboard, chances are it supports it.

When you create a new profile, it saves a name and/or icon... I'm a bit puzzled why it didn't work for you. Unless there is an exception shown when ran in the terminal.

Although, for help with the macros, please do raise an issue on the [driver/daemon repository](https://github.com/terrycain/razer-drivers/issues).

**Comment by terrycain** [comment]
Also device.dedicated_macros should return true signifying that it has actual macro keys

**Comment by dpnova** [comment]
It's RZ03-0039 for the SN. Looks like this http://2f.zol-img.com.cn/product/57/521/ceFp4mQvnlCkE.jpg

See output from the python session below:
```
➜  ~ python3
Python 3.5.2+ (default, Sep 22 2016, 12:18:14) 
[GCC 6.2.0 20160927] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import razer.client
>>> a = razer.client.DeviceManager()
>>> b = a.devices[0]
>>> b.name
'Razer BlackWidow Classic (Alternate)'
>>> b.capabilities
{'lighting_scroll_none': False, 'lighting_logo': False, 'lighting_breath_random': False, 'lighting_breath_triple': False, 'lighting_scroll_breath_random': False, 'lighting_breath_dual': False, 'lighting_breath_single': False, 'lighting_spectrum': False, 'lighting_scroll_static': False, 'lighting_none': False, 'macro_mode_led_effect': True, 'type': True, 'lighting_scroll_spectrum': False, 'lighting_logo_breath_random': False, 'lighting_scroll_breath_single': False, 'lighting_logo_blinking': False, 'lighting_scroll_blinking': False, 'lighting_wave': False, 'lighting_logo_brightness': False, 'lighting_logo_breath_single': False, 'lighting': False, 'macro_logic': True, 'name': True, 'lighting_logo_active': False, 'lighting_ripple_random': False, 'lighting_scroll_breath_dual': False, 'lighting_scroll_reactive': False, 'lighting_logo_spectrum': False, 'brightness': True, 'lighting_scroll_brightness': False, 'lighting_led_single': False, 'lighting_reactive': False, 'firmware_version': True, 'lighting_logo_static': False, 'lighting_scroll': False, 'lighting_led_matrix': False, 'game_mode_led': True, 'lighting_ripple': False, 'serial': True, 'lighting_logo_none': False, 'lighting_logo_pulsate': False, 'macro_tartarus_mode_modifier': False, 'lighting_scroll_pulsate': False, 'lighting_logo_breath_dual': False, 'lighting_logo_reactive': False, 'macro_mode_led': True, 'lighting_static': False, 'lighting_scroll_active': False, 'lighting_pulsate': True}
>>> 
>>> b.dedicated_macro
dbus.Boolean(True)
>>> 
```

**Comment by lah7** [comment]
Thanks for that, Polychromatic will no longer show profiles for devices that do not support the "lighting matrix" (d2bdfc8c4e3d16d87ef43f6f42603f2883a63c50) :+1:

So now it's just a matter of figuring out how to activate macros on your keyboard.

If I run `xev`, pressing FN+F9 (macro key on BW Chroma) recognises the key as this:
```
KeyRelease event, serial 42, synthetic NO, window 0x5e00001,
    root 0x1eb, subw 0x0, time 7146818, (862,-41), root:(863,836),
    state 0x10, keycode 196 (keysym 0x1008ff49, XF86Launch9), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

May be a clue -- in case the daemon is expecting `XF86Launch9` to enter macro mode but it's not the same for your keyboard.

**Comment by terrycain** [comment]
find the device_mode file in /sys/bus/hid/drivers/razerkbd and `cat` it. should output `3:0`

**Comment by dpnova** [comment]
Output from xev:
```
KeyPress event, serial 33, synthetic NO, window 0xc00001,
    root 0x2ac, subw 0x0, time 259023, (456,435), root:(4303,442),
    state 0x0, keycode 108 (keysym 0xffea, Alt_R), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0xc00001,
    root 0x2ac, subw 0x0, time 259141, (456,435), root:(4303,442),
    state 0x8, keycode 108 (keysym 0xffea, Alt_R), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

When i turned my pc on to test this morning, I found that when I press FN - M my macro light is flashing now, but only while I'm holding it down! I've realised this is working before I've started the daemon. Before starting the daemon the combo has no impact on device mode, after the daemon starts, no lights anymore, but it toggles device mode also it seems 1:4 seems to be the device mode immediately after starting the controller. See below

```
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006  
➜  0003:1532:010E.0006 
➜  0003:1532:010E.0006 cat device_mode
3:0
➜  0003:1532:010E.0006 cat device_mode
1:4
➜  0003:1532:010E.0006 cat device_mode
0:0
➜  0003:1532:010E.0006 cat device_mode
3:0
➜  0003:1532:010E.0006 cat device_mode
3:0
➜  0003:1532:010E.0006 cat device_mode
3:0
➜  0003:1532:010E.0006 cat device_mode
3:0
```

I still couldn't get it to record a macro though!

**Comment by dpnova** [comment]
Actually pressing the macro button just registers the two keys individually:

```
KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 678139, (128,275), root:(3975,297),
    state 0x0, keycode 202 (keysym 0x0, NoSymbol), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 678326, (128,275), root:(3975,297),
    state 0x0, keycode 108 (keysym 0xffea, Alt_R), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

**Comment by terrycain** [comment]
Ok in theory when device_mode is 3:0 M1-5 should produce keycodes as well as Fn+F9, is that the case?

**Comment by dpnova** [comment]
Here's me pressing the m1-5 keys with mode `3:0` - I'm pressing return between them so we can tell them apart too. Note as mentioned earlier, only m3-5 actually emit anything and they are they x86 launch ones. The output is the same with mode = `0:0` too with a small exception - I've included one of those below too.

```
FocusOut event, serial 30, synthetic NO, window 0x2400001,
    mode NotifyGrab, detail NotifyAncestor

FocusOut event, serial 30, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyPointer

FocusIn event, serial 30, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyAncestor

KeymapNotify event, serial 30, synthetic NO, window 0x0,
    keys:  4294967265 0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

KeyPress event, serial 30, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 249372, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
"   XmbLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 249423, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

FocusOut event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyGrab, detail NotifyAncestor

FocusOut event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyPointer

FocusIn event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyAncestor

KeymapNotify event, serial 33, synthetic NO, window 0x0,
    keys:  2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 254514, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
"   XmbLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 254585, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 256105, (819,428), root:(826,450),
    state 0x0, keycode 193 (keysym 0x1008ff46, XF86Launch6), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 256188, (819,428), root:(826,450),
    state 0x0, keycode 193 (keysym 0x1008ff46, XF86Launch6), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 257457, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
"   XmbLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 257505, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 263596, (819,428), root:(826,450),
    state 0x0, keycode 194 (keysym 0x1008ff47, XF86Launch7), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 263653, (819,428), root:(826,450),
    state 0x0, keycode 194 (keysym 0x1008ff47, XF86Launch7), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 264058, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
"   XmbLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 264108, (819,428), root:(826,450),
    state 0x0, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyPress event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 264387, (819,428), root:(826,450),
    state 0x0, keycode 195 (keysym 0x1008ff48, XF86Launch8), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

KeyRelease event, serial 33, synthetic NO, window 0x2400001,
    root 0x2ac, subw 0x0, time 264457, (819,428), root:(826,450),
    state 0x0, keycode 195 (keysym 0x1008ff48, XF86Launch8), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```


With device_mode 0:0:

```
FocusOut event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyGrab, detail NotifyAncestor

FocusOut event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyPointer

FocusIn event, serial 33, synthetic NO, window 0x2400001,
    mode NotifyUngrab, detail NotifyAncestor

KeymapNotify event, serial 33, synthetic NO, window 0x0,
    keys:  2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
```

**Comment by dpnova** [comment]
The focus events could be because I'm using xmonad and I have focus follows mouse.. I'm not moving the mouse but maybe the key press is signalling something?

**Comment by terrycain** [comment]
Yeah device mode 3:0 is needed for the M1-5 to emit stuff, Ok that must mean M1 and M2 are doing something proper weird. Possibly because the keyboard is proper old (basing that off the low PID) Its possible M1, M2 and Fn dont emit the same bytes as the Bw chroma and others, that would explain why the daemon isnt doing any macro stuff.

I'd say boot up windows, get a packet capture of pressing m1-5 and fn a couple of times

**Comment by RuRo** [comment]
It seems this thread died? Can I help? I have the BW (classic) [without the Alternate].

**Comment by terrycain** [comment]
^^ looks like driver/daemon related macro stuff is working at least

**Comment by dpnova** [comment]
Yeah I don't have windows to test things in.

I'm very happy to test anything I can on OSX and Linux though.

**Comment by lah7** [comment]
**8 years later:** Polychromatic will be focusing on lighting management. If you need to remap keys, a lot of people use [Input Remapper](https://github.com/sezanzeb/input-remapper) as a utility dedicated to the task.

Key rebinding is pretty generic feature, OpenRazer is more focused on enabling device-specific functionality and lighting. OpenRazer didn't evolve macros beyond the M1-M5 keyboard keys. It's being deprecated and removed in future:

* https://github.com/openrazer/openrazer/pull/2337
